### PR TITLE
Introduction of a detach handler at links in Hono Client.

### PR DESCRIPTION
**Pull request by pellmann**
The client application can react on a detach(close=false) and reopen the link later (retry).
In future this should be handled inside the Hono client as well as the connection retry.

### CODE REVIEW

File: client/src/main/java/org/eclipse/hono/client/HonoClient.java [hunk 1]
```diff
@@ -117,13 +117,48 @@ void getOrCreateEventSender(
...
+
+    /**
+     * Creates a new consumer of telemetry data for a tenant.
+     *
+     * @param tenantId The tenant to consume data for.
+     * @param telemetryConsumer The handler to invoke with every message received.
+     * @param creationHandler The handler to invoke with the outcome of the operation.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    @Deprecated
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/HonoClient.java]: I think we do not need to deprecate this variant. Why not leave it up to clients to choose which variant they want to use? Being forced to pass in `null` to a method invocation never looks good in code FMPOV ...

File: client/src/main/java/org/eclipse/hono/client/HonoClient.java [hunk 2]
```diff
@@ -136,11 +171,31 @@ void createTelemetryConsumer(
      * @param creationHandler The handler to invoke with the outcome of the operation.
      * @throws NullPointerException if any of the parameters is {@code null}.
      */
+    @Deprecated
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/HonoClient.java]: same here

File: client/src/main/java/org/eclipse/hono/client/HonoClient.java [hunk 3]
```diff
@@ -153,6 +208,7 @@ void createEventConsumer(
...
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/HonoClient.java]: same here

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 4]
```diff
@@ -259,19 +267,24 @@ protected static final void setApplicationProperties(final Message msg, final Ma
...
             receiver.openHandler(openAttach -> {
-                if(openAttach.failed()) {
+                if (openAttach.failed()) {
                     LOG.debug("receiver open attach failed [{}] by peer [{}]: {}", receiver.getRemoteSource(), con.getRemoteContainer(), openAttach.cause().getMessage());
                     result.fail(openAttach.cause());
-                }
-                else {
+                } else {
+                    if(LOG.isTraceEnabled()) {
+                        LOG.debug("receiver open attach succeeded [{}] by peer [{}]", receiver.getRemoteSource(), con.getRemoteContainer());
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: do you want to trace or debug?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 5]
```diff
@@ -236,7 +243,8 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             final String sourceAddress,
             final ProtonQoS qos,
             final ProtonMessageHandler messageHandler,
-            final Handler<String> closeHook) {
+            final Handler<String> closeHook,
+            final Handler<AsyncResult<Void>> detachHandler) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: IMHO we should add this as a variant of the existing method with the existing method delegating to the new variant, passing in a `null` detach handler.

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 6]
```diff
@@ -259,19 +267,24 @@ protected static final void setApplicationProperties(final Message msg, final Ma
...
+                        LOG.debug("receiver open attach succeeded [{}] by peer [{}]", receiver.getRemoteSource(), con.getRemoteContainer());
+                    }
                     result.complete(openAttach.result());
                 }
             });
             receiver.detachHandler(remoteDetached -> {
+                receiver.detach();
                 if (remoteDetached.succeeded()) {
                     LOG.debug("receiver [{}] detached (with closed=false) by peer [{}]", receiver.getRemoteSource(), con.getRemoteContainer());
+                    detachHandler.handle(Future.succeededFuture());
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: check if detach handler is `null`

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 7]
```diff
@@ -213,7 +213,7 @@ public final void setRequestTimeout(final long timoutMillis) {
 
     private Future<ProtonReceiver> createReceiver(final ProtonConnection con, final String sourceAddress, final Handler<String> closeHook) {
 
-        return AbstractHonoClient.createReceiver(context, config, con, sourceAddress, ProtonQoS.AT_LEAST_ONCE, this::handleResponse, closeHook);
+        return AbstractHonoClient.createReceiver(context, config, con, sourceAddress, ProtonQoS.AT_LEAST_ONCE, this::handleResponse, closeHook, detach -> {});
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: do not require existing code to pass in *dummy* handlers, instead continue to offer the existing method.

File: client/src/main/java/org/eclipse/hono/client/impl/EventConsumerImpl.java [hunk 8]
```diff
@@ -58,9 +59,10 @@ public static void create(
             final ProtonConnection con,
             final String tenantId,
             final BiConsumer<ProtonDelivery, Message> eventConsumer,
-            final Handler<AsyncResult<MessageConsumer>> creationHandler) {
+            final Handler<AsyncResult<MessageConsumer>> creationHandler,
+            final Handler<AsyncResult<Void>> detachHandler) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/EventConsumerImpl.java]: IMHO we should add this as a variant of the existing method ...

File: client/src/main/java/org/eclipse/hono/client/impl/EventConsumerImpl.java [hunk 9]
```diff
@@ -82,7 +85,8 @@ public static void create(
             final String tenantId,
             final String pathSeparator,
             final BiConsumer<ProtonDelivery, Message> eventConsumer,
-            final Handler<AsyncResult<MessageConsumer>> creationHandler) {
+            final Handler<AsyncResult<MessageConsumer>> creationHandler,
+            final Handler<AsyncResult<Void>> detachHandler) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/EventConsumerImpl.java]: IMHO we should add this as a variant of the existing method ...

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 10]
```diff
@@ -12,20 +12,14 @@
...
-import java.util.function.Supplier;
-
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.proton.ProtonClientOptions;
+import io.vertx.proton.ProtonConnection;
+import io.vertx.proton.ProtonDelivery;
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: please revert this change and adapt your IDE settings to the import order specified in `$HONO/eclipse/hono.importorder`

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 11]
```diff
@@ -41,14 +35,19 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: see above

File: example/src/main/resources/logback.xml [hunk 12]
```diff
@@ -15,6 +15,7 @@
     <appender-ref ref="STDOUT" />
   </root>
 
-  <logger name="org.eclipse.hono.client" level="DEBUG"/>
+  <logger name="org.eclipse.hono.client" level="TRACE"/>
+  <logger name="org.eclipse.hono.client.impl" level="TRACE"/>
```
Comment (sophokles73) [inline example/src/main/resources/logback.xml]: do not commit TRACE log level

File: services/messaging/src/main/java/org/eclipse/hono/messaging/SenderFactoryImpl.java [hunk 13]
```diff
@@ -122,6 +122,14 @@ public void setConfiguration(final HonoMessagingConfigProperties config) {
                 LOG.debug("sender [{}] for container [{}] closed", address, connection.getRemoteContainer());
             }
         });
+        sender.detachHandler(remoteDetached -> {
+            if (remoteDetached.succeeded()) {
+                LOG.debug("sender [{}] detached (with closed=false) by peer [{}]", sender.getRemoteSource(), connection.getRemoteContainer());
+            } else {
+                LOG.debug("sender [{}] detached (with closed=false) by peer [{}]: {}", sender.getRemoteSource(), connection.getRemoteContainer(), remoteDetached.cause().getMessage());
+            }
+            sender.close();
```
Comment (sophokles73) [inline services/messaging/src/main/java/org/eclipse/hono/messaging/SenderFactoryImpl.java]: I do not think that this is enough, we also need to propagate the closing of the downstream link to the upstream protocol adapter. For that, we will probably need to be able to pass in a detach handler...

File: client/src/main/java/org/eclipse/hono/client/HonoClient.java (hunk 1 above)
Comment (pellmann) [reply client/src/main/java/org/eclipse/hono/client/HonoClient.java]: I think the question is if there is a good use case to use the API without this detach handler. My idea was to eliminate the deprecated methods, because without this handlers and a proper reaction there could always be a problem, which could not be healed without this handler. I think we should not provide an API, if we know, the usage will end in problems.

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java (hunk 10 above)
Comment (pellmann) [reply client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: Okay - found the option in the Eclipse Code Formatter plugin for IDEA..

**Comment by pellmann** [comment]
@sophokles73 I merged detach and close hook and also close the upstream receiver link. Could you please take a look.

File: services/messaging/src/main/java/org/eclipse/hono/messaging/ForwardingDownstreamAdapter.java [hunk 14]
```diff
@@ -327,6 +327,15 @@ public final void onClientAttach(final UpstreamReceiver client, final Handler<As
...
                 addSender(client, createdSender);
                 tracker.complete();
+                createdSender.detachHandler(detachHandler -> {
+                    if (detachHandler.succeeded()) {
+                        logger.info("downstream sender was closed detached [con: {}, link: {}]", client.getConnectionId(), client.getLinkId());
+                    } else {
+                        logger.warn("downstream sender was closed detached [con: {}, link: {}]: {}", client.getConnectionId(), client.getLinkId(), detachHandler.cause().getMessage());
+                    }
+                    createdSender.close();
+                    resultHandler.handle(Future.failedFuture(detachHandler.cause()));
```
Comment (sophokles73) [inline services/messaging/src/main/java/org/eclipse/hono/messaging/ForwardingDownstreamAdapter.java]: this seems to be wrong. the resultHandler has already been completed by the *tracker* Future. 
when the peer (the router) detaches this sender link, then we need to do some house keeping:
1. close the link and remove it from the list of downstream senders
2. close the upstream receiver link

Comment (sophokles73) [reply services/messaging/src/main/java/org/eclipse/hono/messaging/ForwardingDownstreamAdapter.java]: I also would prefer if we could pass in a close handler to the `SenderFactory.createSender()` method so that the set up of the link is done in one place only. WDYT?

**Comment by pellmann** [comment]
Will open an other one - conflicts with master.
